{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import kinship\n",
    "import bisect\n",
    "import itertools as iter\n",
    "import phenotype\n",
    "import data_parsers\n",
    "import h5py\n",
    "import numpy\n",
    "import scipy\n",
    "import csv\n",
    "import pdb\n",
    "import itertools\n",
    "from operator import itemgetter\n",
    "from abc import ABCMeta, abstractmethod, abstractproperty\n",
    "\n",
    "log = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "def load_hdf5_genotype_data(hdf5_file):\n",
    "    return HDF5Genotype(hdf5_file)\n",
    "\n",
    "\n",
    "def load_csv_genotype_data(csv_files,format='binary'):\n",
    "    log.info(\"Loading Genotype file\")\n",
    "    snps = []\n",
    "    positions = []\n",
    "    chr_regions = []\n",
    "    chrs = []\n",
    "    accessions = None\n",
    "    if type(csv_files) is list:\n",
    "        log.info(\"Genotype split in %s files\" % len(csv_files))\n",
    "        for i,csv_file in enumerate(csv_files):\n",
    "            log.info(\"Loading %s \" % csv_file)\n",
    "            data = data_parsers.parse_genotype_csv_file(csv_file,format)\n",
    "            if accessions is None:\n",
    "                accessions = data['accessions']\n",
    "            if data['accessions'] != accessions:\n",
    "                raise Exception('Accessions must match')\n",
    "            for chr_data in data['data']:\n",
    "                num_snps = len(chr_data['positions'])\n",
    "                chr_regions.append((len(positions),len(positions)+num_snps))\n",
    "                chrs.append(chr_data['chr'])\n",
    "                positions.extend(chr_data['positions'])\n",
    "                snps.extend(chr_data['snps'])\n",
    "                log.info(\"Finished loading %s SNPs for Chr %s\" % (len(positions),chr_data['chr']))\n",
    "    else:\n",
    "        data = data_parsers.parse_genotype_csv_file(csv_files,format)\n",
    "        accessions = data['accessions']\n",
    "        for chr_data in data['data']:\n",
    "            num_snps = len(chr_data['positions'])\n",
    "            chr_regions.append((len(positions),len(positions)+num_snps))\n",
    "            chrs.append(chr_data['chr'])\n",
    "            positions.extend(chr_data['positions'])\n",
    "            snps.extend(chr_data['snps'])\n",
    "            log.info(\"Finished loading %s SNPs for Chr %s\" % (len(positions),chr_data['chr']))\n",
    "    log.info(\"Finished loading Genotype file: %s SNPs %s accessions %s chromosomes\" % (len(positions),len(accessions),len(chrs)))\n",
    "    return Genotype(snps,positions, accessions,chr_regions,chrs,format)\n",
    "\n",
    "def calculate_ld(snps):\n",
    "        #filter non binary snps\n",
    "        snps_t = scipy.transpose(snps)\n",
    "        snps_stand = scipy.transpose((snps_t - scipy.mean(snps, 1)) / scipy.std(snps, 1))\n",
    "        r2_values =scipy.dot(snps_stand, scipy.transpose(snps_stand))\n",
    "        r2_values *= (1.0 / snps.shape[1])\n",
    "        r2_values **= 2\n",
    "        return r2_values\n",
    "\n",
    "\n",
    "class AbstractGenotype(object):\n",
    "    __metaclass__ = ABCMeta\n",
    "\n",
    "    @abstractproperty\n",
    "    def data_format(self):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def get_snps_iterator(self,chr=None,is_chunked=False,chunk_size=1000):\n",
    "        pass\n",
    "\n",
    "    def get_snp_at(self,chr,position):\n",
    "        chr_ix = chr -1\n",
    "        chr_region = self.chr_regions[chr_ix]\n",
    "        i = bisect.bisect_left(self.positions[chr_region[0]:chr_region[1]], position)\n",
    "        if i != chr_region[1]:\n",
    "            snp_ix = i + chr_region[0]\n",
    "            if self.positions[snp_ix] == position:\n",
    "                return self.snps[snp_ix]\n",
    "        return None\n",
    "\n",
    "    def get_chr_region_ix(self,chr):\n",
    "        # some genotype datasets have chromosomes as integers instead of strings\n",
    "        if self.chrs.dtype.kind == 'i':\n",
    "            return numpy.where(self.chrs == int(chr))[0][0]\n",
    "        return numpy.where(self.chrs == str(chr))[0][0]\n",
    "\n",
    "    @abstractproperty\n",
    "    def positions(self):\n",
    "        pass\n",
    "\n",
    "    @abstractproperty\n",
    "    def accessions(self):\n",
    "        pass\n",
    "\n",
    "    @abstractproperty\n",
    "    def snps(self):\n",
    "        pass\n",
    "\n",
    "    @property\n",
    "    def chromosomes(self):\n",
    "        chromosomes = []\n",
    "        for i,chr_region in enumerate(self.chr_regions):\n",
    "            chromosomes.extend([self.chrs[i]] * (chr_region[1] - chr_region[0]))\n",
    "        return chromosomes\n",
    "\n",
    "    @abstractproperty\n",
    "    def chrs(self):\n",
    "        pass\n",
    "\n",
    "    @abstractproperty\n",
    "    def num_snps(self):\n",
    "        pass\n",
    "\n",
    "    @abstractproperty\n",
    "    def chr_regions(self):\n",
    "        pass\n",
    "\n",
    "    def get_snps_from_pos(self,chr_pos):\n",
    "        \"\"\"\n",
    "        Returns a list of snps based on a list of (chr,pos) tuples\n",
    "        \"\"\"\n",
    "        snps = []\n",
    "        indices = []\n",
    "        if chr_pos is None or len(chr_pos) == 0:\n",
    "            return (indices,snps)\n",
    "        chr_pos_ix = map(list,zip(*sorted(zip(chr_pos,range(len(chr_pos))))))\n",
    "        # group by chr for efficient sequentielly iteration over snps generator\n",
    "        it_ix = 0\n",
    "        filtered_chr_pos_ix = [[],[]]\n",
    "        for chr,positions in iter.groupby(chr_pos_ix[0],lambda x:x[0]):\n",
    "            pos_indices = []\n",
    "            it = self.get_snps_iterator(chr)\n",
    "            for position in positions:\n",
    "                pos_ix = self.get_pos_ix(chr,position[1])\n",
    "                if pos_ix[2] == False:\n",
    "                    it_ix+=1\n",
    "                    continue\n",
    "                filtered_chr_pos_ix[0].append(chr_pos_ix[0][it_ix])\n",
    "                filtered_chr_pos_ix[1].append(chr_pos_ix[1][it_ix])\n",
    "                pos_indices.append(pos_ix[1])\n",
    "                indices.append(pos_ix[0])\n",
    "                it_ix+=1\n",
    "            for i,ix in enumerate(pos_indices):\n",
    "                previous_ix = 0 if i == 0 else pos_indices[i-1] +1\n",
    "                snps.append(next(iter.islice(it,ix-previous_ix,None),None))\n",
    "        return (map(list,zip(*sorted(zip(filtered_chr_pos_ix[1],indices))))[1],map(list,zip(*sorted(zip(filtered_chr_pos_ix[1],snps))))[1])\n",
    "\n",
    "\n",
    "    def get_pos_ix(self,chr,position):\n",
    "        \"\"\"\n",
    "        Returns the index of chr,position using bisect\n",
    "        \"\"\"\n",
    "        chr_region = self.chr_regions[self.get_chr_region_ix(chr)]\n",
    "        positions = self.positions[chr_region[0]:chr_region[1]]\n",
    "        i = bisect.bisect_left(positions, position)\n",
    "        abs_ix = i + chr_region[0]\n",
    "        found = False\n",
    "        if abs_ix != len(self.positions) and self.positions[abs_ix] == position:\n",
    "            found = True\n",
    "        return (abs_ix,i,found)\n",
    "\n",
    "\n",
    "    def get_chr_region_from_index(self,ix):\n",
    "        for i,chr_region in enumerate(self.chr_regions):\n",
    "            if chr_region[0] <= ix and chr_region[1] >= ix:\n",
    "                return i, chr_region\n",
    "        return None\n",
    "\n",
    "    def get_chr_pos_from_index(self,ix):\n",
    "        return (self.get_chr_from_index(ix),self.positions[ix])\n",
    "\n",
    "    def get_chr_from_index(self,ix):\n",
    "        for i,chr_region in enumerate(self.chr_regions):\n",
    "            if chr_region[0] <= ix and chr_region[1] >= ix:\n",
    "                return int(self.chrs[i])\n",
    "        raise Exception('Index %s outside of chr_regions' %ix)\n",
    "\n",
    "    def get_mafs(self):\n",
    "        macs = []\n",
    "        mafs = []\n",
    "        num_nts = len(self.accessions)\n",
    "        if self.data_format in ['binary', 'int']:\n",
    "            for snp in self.get_snps_iterator():\n",
    "                l = scipy.bincount(snp)\n",
    "                mac = min(l)\n",
    "                macs.append(mac)\n",
    "                mafs.append(mac / float(num_nts))\n",
    "        elif self.data_format == 'diploid_int':\n",
    "            for snp in self.get_snps_iterator():\n",
    "                bin_counts = scipy.bincount(snp, minlength=3)\n",
    "                l = scipy.array([bin_counts[0], bin_counts[2]]) + bin_counts[1] / 2.0\n",
    "                mac = l.min()\n",
    "                macs.append(mac)\n",
    "                mafs.append(mac / float(num_nts))\n",
    "        else:\n",
    "            raise NotImplementedError\n",
    "        return {\"macs\":macs, \"mafs\":mafs}\n",
    "\n",
    "    @abstractproperty\n",
    "    def genome_length(self):\n",
    "        pass\n",
    "\n",
    "\n",
    "    def filter_accessions(self,accession_ids):\n",
    "        sd_indices_to_keep = set()\n",
    "        pd_indices_to_keep = []\n",
    "        for i, acc in enumerate(self.accessions):\n",
    "            for j, et in enumerate(accession_ids):\n",
    "                if str(et) == str(acc):\n",
    "                    sd_indices_to_keep.add(i)\n",
    "                    pd_indices_to_keep.append(j)\n",
    "\n",
    "        sd_indices_to_keep = list(sd_indices_to_keep)\n",
    "        sd_indices_to_keep.sort()\n",
    "        self.filter_accessions_ix(sd_indices_to_keep)\n",
    "        return sd_indices_to_keep,pd_indices_to_keep\n",
    "\n",
    "\n",
    "    @abstractmethod\n",
    "    def filter_accessions_ix(self,accessions_ix):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def filter_snps_ix(self,accessions_ix):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def convert_data_format(self,target_format='binary'):\n",
    "        pass\n",
    "\n",
    "    def coordinate_w_phenotype_data(self, phenotype, coord_phen=True):\n",
    "        \"\"\"\n",
    "        Deletes accessions which are not common, and sorts the accessions, removes monomorphic SNPs, etc.\n",
    "        \"\"\"\n",
    "        log.debug(\"Coordinating SNP and Phenotype data.\")\n",
    "        ets = phenotype.ecotypes\n",
    "\n",
    "        #Filter accessions which do not have phenotype values (from the genotype data).\n",
    "        log.debug(\"Filtering accessions\")\n",
    "        sd_indices_to_keep,pd_indices_to_keep =  self.filter_accessions(ets)\n",
    "\n",
    "        if coord_phen:\n",
    "            num_values = phenotype.num_vals\n",
    "            log.debug(\"Filtering phenotype data.\")\n",
    "            #Removing accessions that don't have genotypes or phenotype values\n",
    "            phenotype.filter_ecotypes(pd_indices_to_keep)\n",
    "            ets = phenotype.ecotypes\n",
    "            log.debug(\"Out of %d, leaving %d values.\" % (num_values, len(ets)))\n",
    "        if self.data_format == 'binary':\n",
    "            log.debug('Filtering non-binary SNPs')\n",
    "            (total_num,removed_num) = self.filter_non_binary()\n",
    "            log.debug('Removed %d non-binary SNPs out of %d SNPs' % (removed_num, total_num))\n",
    "        elif self.data_format in ['int', 'diploid_int']:\n",
    "            log.debug('Filtering monomorhpic SNPs')\n",
    "            total_num = 0\n",
    "            removed_num = 0\n",
    "            (total_num,removed_num) = self.filter_monomorphic_snps()\n",
    "            log.debug('Removed %d monomorphic SNPs out of %d SNPs' % (removed_num, total_num))\n",
    "        return {'pd_indices_to_keep':pd_indices_to_keep, 'n_filtered_snps':removed_num}\n",
    "\n",
    "    def get_ibs_kinship_matrix(self, debug_filter=1, snp_dtype='int8', dtype='single',chunk_size=None):\n",
    "        \"\"\"\n",
    "        Calculate the IBS kinship matrix.\n",
    "        (un-scaled)\n",
    "        Currently it works only for binary kinship matrices.\n",
    "        \"\"\"\n",
    "        log.debug('Starting kinship calculation')\n",
    "        return kinship.calc_ibs_kinship(self,chunk_size=chunk_size)\n",
    "\n",
    "    def get_ibd_kinship_matrix(self, debug_filter=1, dtype='single',chunk_size=None):\n",
    "        log.debug('Starting IBD calculation')\n",
    "        return kinship.calc_ibd_kinship(self,chunk_size=chunk_size)\n",
    "        log.debug('Finished calculating IBD kinship matrix')\n",
    "        return cov_mat\n",
    "\n",
    "\n",
    "\n",
    "    def save_as_csv(self,csv_file,chunk_size=1000):\n",
    "        log.info('Writing genotype to CSV file %s' % csv_file)\n",
    "        with open(csv_file,'w') as csvfile:\n",
    "            csv_writer = csv.writer(csvfile,delimiter=',')\n",
    "            header = ['Chromosome','Positions']\n",
    "            header.extend(self.accessions)\n",
    "            csv_writer.writerow(header)\n",
    "            snp_iterator = self.get_snps_iterator(is_chunked=True,chunk_size=chunk_size)\n",
    "            chromosomes = self.chromosomes\n",
    "            positions = self.positions\n",
    "            num = len(positions)\n",
    "            log_iteration_count = num/(chunk_size*10)\n",
    "            for i,snps in enumerate(snp_iterator):\n",
    "                current_ix = i* chunk_size\n",
    "                if i % log_iteration_count == 0:\n",
    "                    log.info('Line %s/%s of Genotype written' % (current_ix,num))\n",
    "                rows = numpy.hstack((numpy.asarray(zip(chromosomes[current_ix:current_ix+chunk_size],positions[current_ix:current_ix+chunk_size])),snps))\n",
    "                csv_writer.writerows(rows.tolist())\n",
    "        log.info('Finished writing genotype ')\n",
    "\n",
    "    def save_as_hdf5(self, hdf5_file):\n",
    "        log.info('Writing genotype to HDF5 file %s' %hdf5_file)\n",
    "        if self.data_format in ['binary', 'diploid_int']:\n",
    "            h5file = h5py.File(hdf5_file, 'w')\n",
    "            num_snps = self.num_snps\n",
    "            num_accessions = len(self.accessions)\n",
    "            h5file.create_dataset('accessions', data=self.accessions,shape=(num_accessions,))\n",
    "            h5file.create_dataset('positions', data=self.positions,shape=(num_snps,),dtype='i4')\n",
    "            h5file['positions'].attrs['chrs'] = self.chrs\n",
    "            h5file['positions'].attrs['chr_regions']  = self.chr_regions\n",
    "            h5file.create_dataset('snps', shape=(num_snps, len(self.accessions)),\n",
    "                              dtype='int8', compression='lzf', chunks=((1000, num_accessions)),data=list(self.get_snps_iterator()))\n",
    "            h5file['snps'].attrs['data_format'] = self.data_format\n",
    "            h5file['snps'].attrs['num_snps'] = num_snps\n",
    "            h5file['snps'].attrs['num_accessions'] = num_accessions\n",
    "            h5file.close()\n",
    "            log.info(\"Finished writing genotype to HDF5 file\")\n",
    "        else:\n",
    "            raise NotImplementedError\n",
    "\n",
    "\n",
    "    def filter_monomorphic_snps(self):\n",
    "        \"\"\"\n",
    "        Removes SNPs from the data which are monomorphic.\n",
    "        \"\"\"\n",
    "        snps_ix = []\n",
    "        num_snps = self.num_snps\n",
    "        for i,snps in enumerate(self.get_snps_iterator()):\n",
    "            bc = numpy.unique(snps)\n",
    "            if len(numpy.unique(snps)) <= 1:\n",
    "                snps_ix.append(i)\n",
    "        numRemoved = len(snps_ix)\n",
    "        self.filter_snps_ix(snps_ix)\n",
    "        log.info(\"Removed %d monomoprhic SNPs, leaving %d SNPs in total.\" % (numRemoved, num_snps))\n",
    "        return (num_snps,numRemoved)\n",
    "\n",
    "\n",
    "    def filter_non_binary(self):\n",
    "        \"\"\"\n",
    "        Removes all but binary SNPs.  (I.e. monomorphic, tertiary and quaternary alleles SNPs are removed.)\n",
    "        \"\"\"\n",
    "        num_snps = self.num_snps\n",
    "        snps_ix = []\n",
    "        num_accessions = len(self.accessions)\n",
    "        # Faster (2.2 ms) than doing numpy.bincount (91.1ms per loop)\n",
    "        # TODO chunk size is hardcoded\n",
    "        for i,snps in enumerate(self.get_snps_iterator(is_chunked=True)):\n",
    "            sm = numpy.sum(snps,axis=1)\n",
    "            snps_ix.extend(numpy.where( (sm == 0) | (sm == num_accessions))[0]+i*1000)\n",
    "        numRemoved = len(snps_ix)\n",
    "        self.filter_snps_ix(snps_ix)\n",
    "        log.info(\"Removed %d non-binary SNPs, leaving %d SNPs in total.\" % (numRemoved, self.num_snps))\n",
    "        return (num_snps,numRemoved)\n",
    "\n",
    "    def calculate_ld(self,chr_pos):\n",
    "        # sort chr_pos first\n",
    "        chr_pos = sorted(chr_pos,key=itemgetter(0,1))\n",
    "        indices,snps = self.get_snps_from_pos(chr_pos)\n",
    "        return calculate_ld(numpy.vstack(snps))\n",
    "\n",
    "\n",
    "class Genotype(AbstractGenotype):\n",
    "\n",
    "    \"\"\"\n",
    "    A class that encompasses multiple _SnpsData_ chromosomes objects (chromosomes), and can deal with them as a whole.\n",
    "    This object should eventually replace the snpsdata lists..\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, snps,positions, accessions,chr_regions, chrs,data_format=None):\n",
    "        self._snps = snps\n",
    "        self._positions = positions\n",
    "        self._chr_regions = chr_regions\n",
    "        self._chrs = chrs\n",
    "        self._accessions = accessions\n",
    "        self._data_format = data_format # binary, diploid_ints, floats, int\n",
    "\n",
    "\n",
    "    @property\n",
    "    def data_format(self):\n",
    "        return self._data_format\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    @property\n",
    "    def accessions(self):\n",
    "        return self._accessions\n",
    "\n",
    "    @property\n",
    "    def snps(self):\n",
    "        return self._snps\n",
    "\n",
    "    @property\n",
    "    def positions(self):\n",
    "        return self._positions\n",
    "\n",
    "    @property\n",
    "    def chrs(self):\n",
    "        return self._chrs\n",
    "\n",
    "    @property\n",
    "    def num_snps(self):\n",
    "        return len(self._snps)\n",
    "\n",
    "    @property\n",
    "    def chr_regions(self):\n",
    "        return self._chr_regions\n",
    "\n",
    "    @property\n",
    "    def genome_length(self):\n",
    "        return self.chr_regions[-1][0]\n",
    "\n",
    "    def filter_snps_ix(self,snps_ix):\n",
    "        self._snps = snps[snps_ix]\n",
    "        self._positions = snps[snps_ix]\n",
    "\n",
    "    def get_snps_iterator(self,chr=None,is_chunked=False,chunk_size=1000):\n",
    "        start = 0\n",
    "        end = None\n",
    "        if chr is not None:\n",
    "            chr_region = self.chr_regions[self.get_chr_region_ix(chr)]\n",
    "            start = chr_region[0]\n",
    "            end = chr_region[1]\n",
    "        if is_chunked:\n",
    "            for i in xrange(start,end,chunk_size):\n",
    "                stop_i = min(i + chunk_size, end)\n",
    "                yield self._snps[i:stop_i]\n",
    "        else:\n",
    "            for snp in self._snps[start:end]:\n",
    "                yield snp\n",
    "\n",
    "\n",
    "    def convert_data_format(self,target_format='binary',reference_ecotype='6909'):\n",
    "        \"\"\"\n",
    "        Converts the underlying raw data format to a binary one, i.e. A,C,G,T,NA,etc. are converted to 0,1,-1\n",
    "        \"\"\"\n",
    "        log.info('Converting from %s to %s' % (self.data_format,target_format))\n",
    "        if self.data_format == target_format:\n",
    "            log.warning(\"Data appears to be already in %s format!\" % target_format)\n",
    "        else:\n",
    "            if self.data_format == 'nucleotides' and target_format == 'binary':\n",
    "                self._convert_snps_to_binary(reference_ecotype = reference_ecotype)\n",
    "            else:\n",
    "                raise NotImplementedError\n",
    "\n",
    "    def _convert_snps_to_binary(self,reference_ecotype='6909'):\n",
    "        missingVal = 'NA'\n",
    "        decoder = {missingVal:-1} #Might cause errors somewhere???!!!\n",
    "        coding_fun = scipy.vectorize(lambda x: decoder[x], otypes=['int8'])\n",
    "\n",
    "        if reference_ecotype in self.accessions:\n",
    "            ref_i = self.accessions.index(reference_ecotype)\n",
    "        else:\n",
    "            ref_i = 0\n",
    "            log.warning(\"Given reference ecotype %s wasn't found, using %s as 0-reference.\" % \\\n",
    "                    (reference_ecotype, self.accessions[ref_i]))\n",
    "        snps = []\n",
    "        num = len(self.positions)\n",
    "        positions = []\n",
    "        chr_region_removed_snps = [0] * len(self.chr_regions)\n",
    "        for snp_i, (snp, pos) in enumerate(itertools.izip(self._snps, self.positions)):\n",
    "            chr_region_ix,chr_region = self.get_chr_region_from_index(snp_i)\n",
    "            if snp_i % (num / 10) == 0:\n",
    "                log.info('Converted %s/%s of SNPs.' % (snp_i,num))\n",
    "            unique_nts = scipy.unique(snp).tolist()\n",
    "            if missingVal in unique_nts:\n",
    "                if len(unique_nts) != 3:\n",
    "                    chr_region_removed_snps[chr_region_ix] +=1\n",
    "                    continue #Skipping non-binary SNP\n",
    "                else:\n",
    "                    unique_nts.remove(self.missingVal)\n",
    "            else:\n",
    "                if len(unique_nts) != 2:\n",
    "                    chr_region_removed_snps[chr_region_ix] +=1\n",
    "                    continue #Skipping non-binary SNP\n",
    "            if snp[ref_i] != missingVal:\n",
    "                ref_nt = snp[ref_i]\n",
    "                decoder[ref_nt] = 0\n",
    "                unique_nts.remove(ref_nt)\n",
    "                decoder[unique_nts[0]] = 1\n",
    "            else:\n",
    "                decoder[unique_nts[0]] = 0\n",
    "                decoder[unique_nts[1]] = 1\n",
    "            snps.append(coding_fun(snp))\n",
    "            positions.append(pos)\n",
    "        log.info('Removed %d non-binary SNPs out of %d, when converting to binary SNPs.'\\\n",
    "            % (len(self.positions) - len(positions), len(self.positions)))\n",
    "        assert len(snps) == len(positions), 'Somthing odd with the lengths.'\n",
    "        chr_regions = self.chr_regions[:]\n",
    "        sum_removed_snps = 0\n",
    "        for i,num_snps in enumerate(chr_region_removed_snps):\n",
    "            sum_removed_snps +=num_snps\n",
    "            if i == 0:\n",
    "                chr_regions[0] = (0,chr_regions[0][1] - num_snps)\n",
    "            else:\n",
    "                chr_regions[i] = (chr_regions[i-1][1],chr_regions[i][1] - sum_removed_snps)\n",
    "        self._chr_regions = chr_regions\n",
    "\n",
    "        self._snps = snps\n",
    "        self._positions = positions\n",
    "        self._data_format = 'binary'\n",
    "\n",
    "\n",
    "    def filter_accessions_ix(self,indicesToKeep):\n",
    "        \"\"\"\n",
    "        Removes accessions from the data.\n",
    "        \"\"\"\n",
    "        num_accessions  = len(self.accessions)\n",
    "        newAccessions = []\n",
    "        newArrayIds = []\n",
    "        for i in indicesToKeep:\n",
    "            newAccessions.append(self.accessions[i])\n",
    "        for i in range(len(self.snps)):\n",
    "            snp = self.snps[i]\n",
    "            newSnp = []\n",
    "            for j in indicesToKeep:\n",
    "                newSnp.append(snp[j])\n",
    "            self.snps[i] = newSnp\n",
    "        self.accessions = newAccessions\n",
    "        # TODO update chr_regions\n",
    "        log.debug(\"Removed %d accessions, leaving %d in total.\" % (num_accessions - len(indicesToKeep), len(indicesToKeep)))\n",
    "\n",
    "\n",
    "class HDF5Genotype(AbstractGenotype):\n",
    "\n",
    "    def __init__(self,hdf5_file):\n",
    "        self.h5file = h5py.File(hdf5_file, 'r')\n",
    "        self.filter_snps = None\n",
    "        self.accession_filter = None\n",
    "\n",
    "    def __del__(self):\n",
    "        if self.h5file is not None:\n",
    "           self.h5file.close()\n",
    "\n",
    "    def get_snps(self):\n",
    "        return self.h5file['snps']\n",
    "\n",
    "    @property\n",
    "    def snps(self):\n",
    "        return self.h5file['snps']\n",
    "\n",
    "    @property\n",
    "    def data_format(self):\n",
    "        return self.h5file['snps'].attrs['data_format']\n",
    "\n",
    "\n",
    "\n",
    "    @property\n",
    "    def accessions(self):\n",
    "        if self.accession_filter is None or len(self.accession_filter) == 0:\n",
    "            return self.h5file['accessions'][:]\n",
    "        else:\n",
    "            return self.h5file['accessions'][self.accession_filter]\n",
    "\n",
    "    def convert_data_format(self,target_format='binary'):\n",
    "        raise NotImplementedError\n",
    "\n",
    "    def _get_snps_(self, start=0,end=None,chunk_size=1000):\n",
    "        \"\"\"\n",
    "        An generator/iterator for SNP chunks.\n",
    "        \"\"\"\n",
    "        if end is None:\n",
    "            end = self.original_num_snps\n",
    "        for i in xrange(start,end,chunk_size):\n",
    "            stop_i = min(i + chunk_size, end)\n",
    "            if self.accession_filter is None or len(self.accession_filter) == 0:\n",
    "                if self.filter_snps is None:\n",
    "                    yield self.h5file['snps'][i:stop_i]\n",
    "                else:\n",
    "                    yield self.h5file['snps'][self.filter_snps[i:stop_i],:]\n",
    "            else:\n",
    "                if self.filter_snps is None:\n",
    "                    # first read the entire row and then filter columns (7.91ms vs 94.3ms)\n",
    "                    yield self.h5file['snps'][i:stop_i][:,self.accession_filter]\n",
    "                else:\n",
    "                    filter_chunk = self.filter_snps[i:stop_i]\n",
    "                    # first read the entire row and then filter columns (7.91ms vs 94.3ms)\n",
    "                    snps_chunk = self.h5file['snps'][i:stop_i][:,self.accession_filter]\n",
    "                    yield snps_chunk[filter_chunk]\n",
    "\n",
    "    def get_snps_iterator(self,chr=None,is_chunked=False,chunk_size=1000):\n",
    "        \"\"\"\n",
    "        Returns an generator containing a chunked generator.\n",
    "        If chr is passed the generator will only iterate over the specified chr\n",
    "        \"\"\"\n",
    "        start = 0\n",
    "        end = None\n",
    "        if chr is not None:\n",
    "\t\t\t# use unfiltered chr_regions because filtering happens in _get_snps_\n",
    "            chr_region = self.h5file['positions'].attrs['chr_regions'][self.get_chr_region_ix(chr)]\n",
    "            start = chr_region[0]\n",
    "            end = chr_region[1]\n",
    "        for snp_chunk in self._get_snps_(start=start,end=end,chunk_size=chunk_size):\n",
    "            if is_chunked:\n",
    "                yield snp_chunk\n",
    "            else:\n",
    "                for snp in snp_chunk:\n",
    "                    yield snp\n",
    "\n",
    "    @property\n",
    "    def original_num_snps(self):\n",
    "        return self.h5file['snps'].attrs['num_snps']\n",
    "\n",
    "    @property\n",
    "    def positions(self):\n",
    "        if self.filter_snps is not None:\n",
    "            return self.h5file['positions'][self.filter_snps]\n",
    "        return self.h5file['positions'][:]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    @property\n",
    "    def chrs(self):\n",
    "        return self.h5file['positions'].attrs['chrs']\n",
    "\n",
    "    @property\n",
    "    def num_snps(self):\n",
    "        if self.filter_snps is not None:\n",
    "            return self.filter_snps.sum()\n",
    "        return self.original_num_snps\n",
    "\n",
    "    @property\n",
    "    def chr_regions(self):\n",
    "        if self.filter_snps is not None:\n",
    "            return self.filtered_chr_regions\n",
    "        return self.h5file['positions'].attrs['chr_regions']\n",
    "\n",
    "    @property\n",
    "    def genome_length(self):\n",
    "        return self.chr_regions[-1][0]\n",
    "\n",
    "\n",
    "    def filter_accessions_ix(self,indicesToKeep):\n",
    "        \"\"\"\n",
    "        Removes accessions from the data.\n",
    "        \"\"\"\n",
    "        num_accessions = len(self.accessions)\n",
    "        self.accession_filter = indicesToKeep\n",
    "        log.debug(\"Removed %d accessions, leaving %d in total.\" % (num_accessions - len(indicesToKeep), len(indicesToKeep)))\n",
    "\n",
    "\n",
    "    def filter_snps_ix(self,snps_ix):\n",
    "        if snps_ix is None or len(snps_ix) == 0:\n",
    "            self.filter_snps = None\n",
    "            self.filtered_chr_regions = None\n",
    "        else:\n",
    "            self.filter_snps = numpy.ones((self.num_snps,),dtype=numpy.bool)\n",
    "            self.filter_snps[snps_ix] = 0\n",
    "            self.filtered_chr_regions = self._get_filtered_regons()\n",
    "\n",
    "\n",
    "    def _get_filtered_regons(self):\n",
    "        filtered_chr_regions = []\n",
    "        if self.filter_snps is None:\n",
    "            return None\n",
    "        start_ix = 0\n",
    "        end_ix = 0\n",
    "        for chr_region in self.h5file['positions'].attrs['chr_regions']:\n",
    "            end_ix = start_ix + self.filter_snps[chr_region[0]:chr_region[1]].sum()\n",
    "            filtered_chr_regions.append((start_ix,end_ix))\n",
    "            start_ix = end_ix\n",
    "        return filtered_chr_regions\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
